# Comparison pipeline
# psudorule "final" or "all" should specify explicitly which files to create!
# Links: http://watson.nci.nih.gov/~sdavis/blog/flexible_bioinformatics_pipelines_with_snakemake/
# https://github.com/inodb/snakemake-uppmax-demo

rule targets:
  message: 'Pseudo rule'
  input: 'analysis/completed/000134T.completed'

# +--------------------------------------------------------------------+
# | Sequencing variants preprocessing
# +--------------------------------------------------------------------+
rule extract:
  message: 'Extract variants based on RS numbers and ensure synced sorting'
  input: '../VARIANT/samples/{sample}.vcf'
  output: 'sequencing/{sample}.slim.vcf'
  shell: """
    taboo extract references/rsnumbers.txt {input} \
      | taboo filter \
      | vcf-sort \
      > {output}
  """

# +--------------------------------------------------------------------+
# | Genotype comparison
# +--------------------------------------------------------------------+
rule compare:
  message: 'Compare genotype between VCFs from the same sample'
  input: seq='sequencing/{sample}.slim.vcf', maf='maf/samples/{sample}.vcf'
  output: 'analysis/{sample}.tsv'
  shell: 'taboo compare {input.seq} {input.maf} > {output}'

# +--------------------------------------------------------------------+
# | Load results into database
# +--------------------------------------------------------------------+
rule load:
  message: 'Load results into the database'
  input: rules.compare.output
  output: 'analysis/completed/{sample}.completed'
  shell: """
    taboo load --uri=./temp-test.sqlite3 {input} && \
    touch {output}
  """
